controllers/csiaddonsnode_controller.go

```diff
@@ -39,17 +50,98 @@ type CSIAddonsNodeReconciler struct {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the CSIAddonsNode object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *CSIAddonsNodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	logger.Info("Starting Reconcile")
+	// Fetch CSIAddonsNode instance
+	instance := &csiaddonsv1alpha1.CSIAddonsNode{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("CSIAddonsNode resource not found")
+
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	nodeID := instance.Spec.Driver.NodeID
+	driverName := instance.Spec.Driver.Name
+	endPoint := instance.Spec.Driver.EndPoint
+	key := r.creatKey(instance.Namespace, instance.Name)
+
+	logger = logger.WithValues(
+		"NodeID", nodeID,
+		"EndPoint", endPoint,
+		"Driver Name", driverName)
+
+	if !instance.DeletionTimestamp.IsZero() {
+		logger.Info("Deleting connection")
+		r.ConnPool.Delete(key)
+
+		if util.ContainsInSlice(instance.Finalizers, csiAddonsNodeFinalizer) {
+			logger.Info("Removing finalizer")
+
+			instance.Finalizers = util.RemoveFromSlice(instance.Finalizers, csiAddonsNodeFinalizer)
+			if err = r.Client.Update(ctx, instance); err != nil {
+				logger.Error(err, "Failed to remove finalizer")
+
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("Exiting Reconcile")
 
-	// TODO(user): your logic here
+		return ctrl.Result{}, nil
+	}
+
+	if !util.ContainsInSlice(instance.Finalizers, csiAddonsNodeFinalizer) {
+		logger.Info("Adding finalizer")
+
+		instance.Finalizers = append(instance.Finalizers, csiAddonsNodeFinalizer)
+		if err = r.Client.Update(ctx, instance); err != nil {
+			logger.Error(err, "Failed to add finalizer")
+
+			return ctrl.Result{}, err
+		}
+	}
+
+	logger.Info("Connecting to sidecar")
+	newCon, err := connection.NewConnection(
+		ctx,
+		endPoint,
+		nodeID,
+		driverName)
+	if err != nil {
+		s, _ := status.FromError(err)
+		logger.Error(s.Err(), "Failed to establish connection with sidecar")
+
+		instance.Status.State = csiaddonsv1alpha1.CSIAddonsNodeStateFailed
+		instance.Status.Message = fmt.Sprintf("Failed to establish connection with sidecar: %v", s.Message())
```

Comment:
s.Message exists only if `status.FromError` can convert the error to GRPC. can you check the bool value from `status.FromError` and access s.Message?